server/service/sessions.go

```diff
@@ -210,6 +240,65 @@ func (svc *Service) Login(ctx context.Context, email, password string) (*fleet.U
 	return user, session, nil
 }
 
+func (svc *Service) makeSession(ctx context.Context, userID uint) (*fleet.Session, error) {
+	return svc.ds.NewSession(ctx, userID, svc.config.Session.KeySize)
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Session create (second step of MFA)
+////////////////////////////////////////////////////////////////////////////////
+
+type sessionCreateRequest struct {
+	Token string `json:"token,omitempty"`
+}
+
+func sessionCreateEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
+	req := request.(*sessionCreateRequest)
+	session, user, err := svc.CompleteMFA(ctx, req.Token)
+	if err != nil {
+		return loginResponse{Err: err}, nil
+	}
+	// Add viewer to context to allow access to service teams for list of available teams.
+	ctx = viewer.NewContext(ctx, viewer.Viewer{
+		User:    user,
+		Session: session,
+	})
+	availableTeams, err := svc.ListAvailableTeamsForUser(ctx, user)
+	if err != nil {
+		if errors.Is(err, fleet.ErrMissingLicense) {
+			availableTeams = []*fleet.TeamSummary{}
+		} else {
+			return loginResponse{Err: err}, nil
+		}
+	}
+	return loginResponse{user, availableTeams, session.Key, nil}, nil
+}
+
+func (svc *Service) CompleteMFA(ctx context.Context, token string) (*fleet.Session, *fleet.User, error) {
+	// skipauth: No user context available yet to authorize against.
+	svc.authz.SkipAuthorization(ctx)
+
+	var err error
+	defer func(start time.Time) { // force MFA failures to take at least a second for brute force/timing attack resistance
+		if err != nil {
+			time.Sleep(time.Until(start.Add(1 * time.Second)))
+		}
+	}(time.Now())
+
+	session, user, err := svc.ds.SessionByMFAToken(ctx, token, svc.config.Session.KeySize)
+	if err != nil {
+		return nil, nil, fleet.NewAuthFailedError(err.Error())
+	}
+
+	if err := svc.NewActivity(
```

Comment:
Thanks, missed that. Will fix.

Reply:
Actually, looks like that was intentional on the other places that did this, since if we failed to create the success activity then we fail closed but don't need to do timing attack prevention because the user should've been able to log in.